added method download mod (tarball) from github

BRO3886 opened the pull request:
* generates a 'modules' folder
* inside the modules folder, it separates by source, such as "api.github.com"
* inside the source folder there's separation for user/repo

* downloads the tarball
* logs an error if unable to download

* added access token example as well (sample.env)

BRO3886 committed:
added function to download tar from repo

BRO3886 committed:
added env sample

SuperFola commented on .env.sample:
```diff
@@ -0,0 +1 @@
+GITHUB_ACCESS_TOKEN=<YOUR ACCESS TOKEN>
```
I went with this kind of strategy (I forgot to push it long ago, and wondered where it was when I reviewed your PR) https://github.com/ArkScript-lang/nuclear/blob/master/nuclear/__main__.py#L67-L84, maybe it's not the way to go and .env files are better?

BRO3886 replied:
this is a sample .env for people to clone and get started with, I am using a .env file only

SuperFola replied:
no problem, let's ditch the nuclear config file then

SuperFola commented on nuclear/install.py:
```diff
@@ -31,6 +36,13 @@ def handle(args: argparse.Namespace) -> int:
+        filename = get_dir.get_filename(r.headers.get('content-disposition'))
+        try:  
+            makedirs(get_dir.get_dir_name(tar_addr))
+            open(get_dir.get_dir_name(tar_addr)+"/"+filename,'xb').write(r.content)
```
maybe do it using `with` so that the opened file get closed at the end

BRO3886 replied:
Okay I'll do this

SuperFola commented:
About that part:

> * generates a 'modules' folder
> * inside the modules folder, it separates by source, such as "api.github.com"

The package manager shouldn't work like node, nor like pip, but install the packages in a central "local repository" (somewhere like `~/ark-pkgs` for example, which could be left to be configured by the end user, referenced in a sort of packages.json per project).

The idea to separate per source and per user is great, I would also add that the package manager was designed to be able to handle multiple version of the same package, so it would need an organization like `source/user/repo/version`.

Later, each installed module will be checked for any "packages.json" file (or whatever we'll call it) to install its dependencies (which shall get listed as well under the "packages.json" of the current project), so that the module loader can read this file and get the exact wanted file.

BRO3886 commented:
> About that part:
> 
> > * generates a 'modules' folder
> > * inside the modules folder, it separates by source, such as "api.github.com"
> 
> The package manager shouldn't work like node, nor like pip, but install the packages in a central "local repository" (somewhere like `~/ark-pkgs` for example, which could be left to be configured by the end user, referenced in a sort of packages.json per project).
> 
> The idea to separate per source and per user is great, I would also add that the package manager was designed to be able to handle multiple version of the same package, so it would need an organization like `source/user/repo/version`.
> 
> Later, each installed module will be checked for any "packages.json" file (or whatever we'll call it) to install its dependencies (which shall get listed as well under the "packages.json" of the current project), so that the module loader can read this file and get the exact wanted file.


Can you send some package links for me to work with? It'll help with the version-folder creation

BRO3886 committed:
used with open function to close file at the end of file write

BRO3886 committed:
now it downloads all packages to the home root directory in a folder called arkscript-pkgs

BRO3886 committed:
added instructions for troubleshooting github rate limiting and generating oauth token

SuperFola commented on README.md:
````diff
@@ -29,6 +29,22 @@ An ArkScript Package Manager
+GITHUB_ACCESS_TOKEN=<YOUR ACCESS TOKEN>
+```
+* save and exit
+* Checkout [sample.env](./sample.env) to see how the .env file should look like
````
.env.sample

SuperFola commented:
Also, could you get up to date with master?

----

As for the tests, alas we don't have any repositories, I'm just testing with ArkScript-lang/std (I will add a 0.0.1 release since we finally have documentation for all of the functions).

BRO3886 committed:
*commented out the authentication group
*edited the get.py get function to take github access token from .env, if present
*edited readme

BRO3886 committed:
resolve conf

BRO3886 committed:
updated the authentication group to take just token, which is stored in .env

BRO3886 committed:
update readme

BRO3886 committed:
github token bugfix

BRO3886 committed:
github token bugfix

BRO3886 committed:
support version-ed downloads

BRO3886 committed:
edited inline comments

BRO3886 commented:
Solves #1 and #3 I guess

SuperFola reviewed (commented):
Seems good to me, I'll test that tomorrow and it should be merged!

Thanks for your contribution!

SuperFola committed:
Merge branch 'master' into master

BRO3886 commented:
https://hacktoberfest.digitalocean.com/hacktoberfest-update?updated

Can you add the appropriate tags please?

SuperFola commented:
Hey! AFAIK, no need to add a topic to the PR since it's merged and the repo is correctly labelled!
